File: tests/ui/crud/delete-share.js
const {chromium} = require('playwright')
const path = require('path')
const {login, getConfig, getWalletsFromStorage} = require('../auth-helper')

/**
 * Test: Delete wallet share (revoke access)
 * This test deletes/revokes an existing wallet share
 */

;(async () => {
  const browser = await chromium.launch({headless: true, slowMo: 500})
  const page = await browser.newPage()
  const config = getConfig()

  let success = false

  try {
    console.log('🚀 Starting delete wallet share test...')

    // Monitor DELETE requests
    let lastDeleteResponse = null
    page.on('response', response => {
      if (
        response.url().includes('/api/v1/wallet_shares/') &&
        response.request().method() === 'DELETE'
      ) {
        lastDeleteResponse = {status: response.status(), url: response.url()}
        console.log(
          `📥 DELETE response: ${response.status()} ${response.url()}`
        )
      }
    })

    // Step 1: Login as admin
    console.log('📝 Step 1: Logging in as admin...')
    await login(page)

    // Step 2: Check user has a wallet via UI
    console.log('📝 Step 2: Checking user has a wallet...')
    await page.goto(`${config.baseUrl}/wallet`)
    await page.waitForTimeout(2000)

    // Count wallet items in UI
    const walletCount = await page
      .locator('.q-drawer .q-list .q-item')
      .filter({hasNot: page.locator('text=Add a new wallet')})
      .count()

    if (walletCount === 0) {
      throw new Error(
        'No wallets found in UI. Please run create-wallet.js first.'
      )
    }

    console.log(`✅ User has ${walletCount} wallet(s)`)

    // Get wallet data from localStorage for API calls
    const wallets = await getWalletsFromStorage(page)
    const existingWallet = wallets[0]
    const walletId = existingWallet.id
    const adminKey = existingWallet.adminkey

    console.log(`✅ Using wallet ID: ${walletId}`)

    // Step 3: Get existing shares
    console.log('📝 Step 3: Getting existing shares...')
    const response = await page.request.get(
      `${config.baseUrl}/api/v1/wallet_shares/${walletId}`,
      {
        headers: {'X-Api-Key': adminKey}
      }
    )

    if (!response.ok()) {
      throw new Error(`Failed to get shares: ${response.status()}`)
    }

    let shares = await response.json()
    let initialCount = Array.isArray(shares) ? shares.length : 0

    console.log(`📊 Initial share count: ${initialCount}`)

    // If no shares exist, create one first
    if (initialCount === 0) {
      console.log('⚠️ No shares found. Creating one first...')
      const shareWithUser = config.secondaryUsername
      if (!shareWithUser) {
        throw new Error('LNBITS_SECONDARY_USERNAME must be set in .env.local')
      }

      const createResponse = await page.request.post(
        `${config.baseUrl}/api/v1/wallet_shares/${walletId}`,
        {
          headers: {
            'X-Api-Key': adminKey,
            'Content-Type': 'application/json'
          },
          data: JSON.stringify({
            user_id: shareWithUser,
            permissions: 1
          })
        }
      )

      if (!createResponse.ok()) {
        throw new Error(`Failed to create share: ${createResponse.status()}`)
      }

      console.log('✅ Created test share')

      // Re-fetch shares
      const refetchResponse = await page.request.get(
        `${config.baseUrl}/api/v1/wallet_shares/${walletId}`,
        {
          headers: {'X-Api-Key': adminKey}
        }
      )
      shares = await refetchResponse.json()
      initialCount = Array.isArray(shares) ? shares.length : 0
    }

    const shareToDelete = shares[0]
    console.log(`📄 Found share to delete: ${shareToDelete.id}`)
    console.log(`   User: ${shareToDelete.username || shareToDelete.user_id}`)

    // Step 4: Delete share via API
    console.log('📝 Step 4: Deleting share...')

    const deleteResponse = await page.request.delete(
      `${config.baseUrl}/api/v1/wallet_shares/${shareToDelete.id}`,
      {
        headers: {'X-Api-Key': adminKey}
      }
    )

    if (deleteResponse.ok()) {
      console.log(`✅ Share deleted successfully`)
      success = true
    } else {
      console.log(`❌ Delete failed: ${deleteResponse.status()}`)
      const errorText = await deleteResponse.text()
      console.log(`   Error: ${errorText}`)
    }

    // Step 5: Verify deletion
    console.log('📝 Step 5: Verifying deletion...')
    await page.waitForTimeout(2000)

    const verifyResponse = await page.request.get(
      `${config.baseUrl}/api/v1/wallet_shares/${walletId}`,
      {
        headers: {'X-Api-Key': adminKey}
      }
    )

    let finalCount = 0
    if (verifyResponse.ok()) {
      const updatedShares = await verifyResponse.json()
      finalCount = Array.isArray(updatedShares) ? updatedShares.length : 0

      console.log(`📊 Final share count: ${finalCount}`)
      console.log(`📉 Count change: ${finalCount - initialCount}`)

      // Verify the specific share status changed to 'revoked'
      const updatedShare = updatedShares.find(s => s.id === shareToDelete.id)

      if (updatedShare && updatedShare.status === 'revoked') {
        console.log('✅ Deletion verified - share status is now revoked!')
        success = true
      } else if (!updatedShare) {
        console.log('✅ Deletion verified - share no longer exists!')
        success = true
      } else {
        console.log(
          `❌ Deletion verification failed - status is ${updatedShare.status}, expected 'revoked'`
        )
        success = false
      }
    }

    // Step 6: Test UI deletion (optional - if we have more shares)
    if (finalCount > 0) {
      console.log('\n📝 Step 6: Testing UI deletion...')

      // Open Share Wallet dialog
      // The Share Wallet button is a round button with group icon in top right
      const shareButton = page
        .locator('button.text-deep-purple i.material-icons:has-text("group")')
        .locator('..')
      if (await shareButton.isVisible({timeout: 5000})) {
        await shareButton.click()
        await page.waitForTimeout(2000)

        // Take screenshot of shares before deletion
        await page.screenshot({
          path: path.join(__dirname, '../screenshots/delete-share-before.png'),
          fullPage: true
        })

        // Look for delete/revoke button (trash icon or similar)
        const deleteButtons = page.locator(
          'button[aria-label="Delete"], button:has(i:has-text("delete")), .q-btn:has(.q-icon:has-text("delete"))'
        )
        const deleteButtonCount = await deleteButtons.count()

        if (deleteButtonCount > 0) {
          console.log(`🗑️  Found ${deleteButtonCount} delete button(s)`)

          // Click the first delete button
          await deleteButtons.first().click()
          await page.waitForTimeout(1000)

          // Handle confirmation dialog if present
          const confirmButton = page.locator(
            'button:has-text("OK"), button:has-text("Confirm"), button:has-text("Yes")'
          )
          if (await confirmButton.isVisible({timeout: 2000})) {
            await confirmButton.click()
            console.log('✅ Confirmed deletion')
          }

          await page.waitForTimeout(2000)

          // Take screenshot after UI deletion
          await page.screenshot({
            path: path.join(__dirname, '../screenshots/delete-share-after.png'),
            fullPage: true
          })

          console.log('✅ UI deletion test complete')
        } else {
          console.log('⚠️ No delete buttons found in UI')
        }
      }
    }

    // Take final screenshot
    await page.screenshot({
      path: path.join(__dirname, '../screenshots/delete-share-result.png'),
      fullPage: true
    })

    if (success) {
      console.log('\n🎉 SUCCESS! Share deleted successfully!')
    }
  } catch (error) {
    console.error('💥 Error:', error.message)
    await page.screenshot({
      path: path.join(__dirname, '../screenshots/delete-share-error.png'),
      fullPage: true
    })
  } finally {
    await browser.close()
  }

  process.exit(success ? 0 : 1)
})()

File: tests/api/delete_share.py
#!/usr/bin/env python3
"""
Test: Delete wallet share via API
This test deletes/revokes an existing wallet share using the REST API
"""

import asyncio
import os

import httpx
from loguru import logger


# Load config from .env.local
def load_config():
    config = {}
    env_path = os.path.join(os.path.dirname(__file__), "../../.env.local")

    if os.path.exists(env_path):
        with open(env_path) as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#") and "=" in line:
                    key, value = line.split("=", 1)
                    config[key.strip()] = value.strip()

    return {
        "base_url": config.get("TEST_LNBITS_URL", os.getenv("TEST_LNBITS_URL")),
        "admin_key": config.get("TEST_ADMIN_API_KEY", os.getenv("TEST_ADMIN_API_KEY")),
        "wallet_id": config.get("TEST_WALLET_ID", os.getenv("TEST_WALLET_ID")),
        "username": config.get(
            "LNBITS_ADMIN_USERNAME", os.getenv("LNBITS_ADMIN_USERNAME")
        ),
        "password": config.get(
            "LNBITS_ADMIN_PASSWORD", os.getenv("LNBITS_ADMIN_PASSWORD")
        ),
    }


async def test_delete_share():  # noqa: C901
    """Test deleting wallet share"""
    config = load_config()

    if not config["admin_key"]:
        logger.error("❌ TEST_ADMIN_API_KEY must be set in .env.local")
        return False

    if not config["wallet_id"]:
        logger.error("❌ TEST_WALLET_ID must be set in .env.local")
        return False

    logger.info("🚀 Starting delete wallet share API test...")
    logger.info(f"📍 Base URL: {config['base_url']}")
    logger.info(f"📍 Wallet ID: {config['wallet_id']}")

    async with httpx.AsyncClient() as client:
        # Step 1: Get existing shares
        logger.info("📝 Step 1: Getting existing shares...")
        response = await client.get(
            f"{config['base_url']}/api/v1/wallet_shares/{config['wallet_id']}",
            headers={"X-Api-Key": config["admin_key"]},
        )

        if response.status_code != 200:
            logger.error(f"❌ Failed to get shares: {response.status_code}")
            return False

        shares = response.json()
        initial_count = len(shares) if isinstance(shares, list) else 0

        logger.info(f"📊 Initial share count: {initial_count}")

        if initial_count == 0:
            logger.warning("⚠️  No shares found to delete. Please create a share first.")
            return True  # Not a failure, just nothing to delete

        share_to_delete = shares[0]
        logger.info(f"📄 Found share to delete: {share_to_delete['id']}")
        logger.info(f"   User: {share_to_delete.get('username', 'Unknown')}")

        # Security check: Verify user_id is NOT exposed
        if "user_id" in share_to_delete:
            logger.error(
                "❌ SECURITY ISSUE: user_id should not be exposed in API response!"
            )
            return False

        # Step 2: Delete share (using admin key)
        logger.info("📝 Step 2: Deleting share...")
        response = await client.delete(
            f"{config['base_url']}/api/v1/wallet_shares/{share_to_delete['id']}",
            headers={"X-Api-Key": config["admin_key"]},
        )

        if response.status_code != 200:
            logger.error(f"❌ Failed to delete share: {response.status_code}")
            logger.error(f"   Response: {response.text}")
            return False

        logger.info("✅ Share deleted successfully!")

        # Step 3: Verify deletion
        logger.info("📝 Step 3: Verifying deletion...")
        response = await client.get(
            f"{config['base_url']}/api/v1/wallet_shares/{config['wallet_id']}",
            headers={"X-Api-Key": config["admin_key"]},
        )

        if response.status_code != 200:
            logger.error(f"❌ Failed to verify: {response.status_code}")
            return False

        updated_shares = response.json()
        final_count = len(updated_shares) if isinstance(updated_shares, list) else 0

        logger.info(f"📊 Final share count: {final_count}")
        logger.info(f"📉 Count change: {final_count - initial_count}")

        # Verify the specific share is gone
        still_exists = any(s["id"] == share_to_delete["id"] for s in updated_shares)

        if not still_exists and final_count == initial_count - 1:
            logger.info("✅ Deletion verified - share no longer exists!")
            logger.info("\n🎉 SUCCESS! Share deleted successfully!")
            return True
        else:
            logger.error("❌ Deletion verification failed")
            if still_exists:
                logger.error("   Share still exists after deletion")
            if final_count != initial_count - 1:
                logger.error(
                    f"   Expected count: {initial_count - 1}, Got: {final_count}"
                )
            return False


if __name__ == "__main__":
    success = asyncio.run(test_delete_share())
    exit(0 if success else 1)
